src/dsa/rpo_falcon512/keys/secret_key.rs

```diff
@@ -0,0 +1,327 @@
+    pub fn from_seed(seed: [u8; 32]) -> Self {
+        let basis = ntru_gen(N, seed);
+        Self::from_short_lattice_basis(basis)
+    }
+
+    /// Given a short basis [[g, -f], [G, -F]], computes the normalized LDL tree i.e., Falcon tree.
+    fn from_short_lattice_basis(basis: ShortLatticeBasis) -> SecretKey {
+        // FFT each polynomial of the short basis.
+        let basis_fft = to_complex_fft(&basis);
+        // compute the Gram matrix.
+        let gram_fft = gram(basis_fft);
+        // construct the LDL tree of the Gram matrix.
+        let mut tree = ffldl(gram_fft);
+        // normalize the leaves of the LDL tree.
+        normalize_tree(&mut tree, SIGMA);
+        Self { secret_key: basis, tree }
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns the polynomials of the short lattice basis of this secret key.
+    pub fn short_lattice_basis(&self) -> &ShortLatticeBasis {
+        &self.secret_key
+    }
+
+    /// Returns the public key corresponding to this secret key.
+    pub fn public_key(&self) -> PublicKey {
+        self.compute_pub_key_poly().into()
+    }
+
+    /// Returns the LDL tree associated to this secret key.
+    pub fn tree(&self) -> &LdlTree {
+        &self.tree
+    }
+
+    // SIGNATURE GENERATION
+    // --------------------------------------------------------------------------------------------
+
+    /// Signs a message with the secret key.
+    ///
+    /// Takes a randomness generator implementing `Rng` and outputs a signature `Signature`.
+    ///
+    /// # Errors
+    /// Returns an error of signature generation fails.
+    pub fn sign<R: Rng>(&self, message: Word, rng: &mut R) -> Result<Signature, FalconError> {
+        let mut nonce_bytes = [0u8; SIG_NONCE_LEN];
+        rng.fill_bytes(&mut nonce_bytes);
+        let nonce = Nonce::new(nonce_bytes);
+
+        let c = hash_to_point_rpo256(message, &nonce);
+        let (s1, s2) = self.sign_helper(c, rng)?;
+
+        Ok(Signature::new(nonce, s1, s2))
+    }
+
+    // HELPER METHODS
+    // --------------------------------------------------------------------------------------------
+
+    /// Derives the public key corresponding to this secret key using h = g /f [mod ϕ][mod p].
+    fn compute_pub_key_poly(&self) -> PubKeyPoly {
+        let g: Polynomial<FalconFelt> = self.secret_key[0].clone().into();
+        let g_fft = g.fft();
+        let minus_f: Polynomial<FalconFelt> = self.secret_key[1].clone().into();
+        let f = -minus_f;
+        let f_fft = f.fft();
+        let h_fft = g_fft.hadamard_div(&f_fft);
+        h_fft.ifft().into()
+    }
+
+    /// Signs a message polynomial with the secret key.
+    ///
+    /// Takes a randomness generator implementing `Rng` and message polynomial representing `c`
+    /// the hash-to-point of the message to be signed. It outputs a signature `Signature`.
```

Comment:
We should update this comment to describe the correct return type.

Reply:
Updated